### PR TITLE
Fixed #32375 -- Started deprecation toward changing the default sitemap protocol to https.

### PR DESCRIPTION
**Pull request by rohithpr**
https://code.djangoproject.com/ticket/32375

### CODE REVIEW

File: django/contrib/sitemaps/__init__.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import warnings
```
Comment (rohithpr) [inline django/contrib/sitemaps/__init__.py:1]: I'm not sure if the title should say "Fixed" as this PR only deprecates http as the default protocol but doesn't actually change the protocol to https. Please let me know if I should be making any changes.

File: django/contrib/sitemaps/__init__.py [hunk 2]
```diff
@@ -7,6 +8,7 @@
 from django.core.exceptions import ImproperlyConfigured
 from django.urls import NoReverseMatch, reverse
 from django.utils import translation
+from django.utils.deprecation import RemovedInDjango41Warning
```
Comment (pope1ni) [inline django/contrib/sitemaps/__init__.py:11]:
```suggestion
from django.utils.deprecation import RemovedInDjango50Warning
```

File: tests/sitemaps_tests/test_generic.py [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 from django.contrib.sitemaps import GenericSitemap
 from django.test import override_settings
+from django.utils.deprecation import RemovedInDjango41Warning
```
Comment (pope1ni) [inline tests/sitemaps_tests/test_generic.py:5]:
```suggestion
from django.utils.deprecation import RemovedInDjango50Warning
```

File: tests/sitemaps_tests/test_http.py [hunk 4]
```diff
@@ -4,8 +4,9 @@
 from django.contrib.sitemaps import Sitemap
 from django.contrib.sites.models import Site
 from django.core.exceptions import ImproperlyConfigured
-from django.test import modify_settings, override_settings
+from django.test import ignore_warnings, modify_settings, override_settings
 from django.utils import translation
+from django.utils.deprecation import RemovedInDjango41Warning
```
Comment (pope1ni) [inline tests/sitemaps_tests/test_http.py:9]:
```suggestion
from django.utils.deprecation import RemovedInDjango50Warning
```

File: tests/sitemaps_tests/test_http.py [hunk 5]
```diff
@@ -197,6 +198,7 @@ def test_requestsite_sitemap(self):
 """ % date.today()
         self.assertXMLEqual(response.content.decode(), expected_content)
 
+    @ignore_warnings(category=RemovedInDjango41Warning)
```
Comment (pope1ni) [inline tests/sitemaps_tests/test_http.py:201]:
```suggestion
    @ignore_warnings(category=RemovedInDjango50Warning)
```

File: tests/sitemaps_tests/test_http.py [hunk 6]
```diff
@@ -216,6 +219,7 @@ def test_sitemap_get_urls_no_site_2(self):
         with self.assertRaisesMessage(ImproperlyConfigured, self.use_sitemap_err_msg):
             Sitemap().get_urls()
 
+    @ignore_warnings(category=RemovedInDjango41Warning)
```
Comment (pope1ni) [inline tests/sitemaps_tests/test_http.py:222]:
```suggestion
    @ignore_warnings(category=RemovedInDjango50Warning)
```

File: tests/sitemaps_tests/test_http.py [hunk 7]
```diff
@@ -207,6 +209,7 @@ def test_sitemap_get_urls_no_site_1(self):
             Sitemap().get_urls()
 
     @modify_settings(INSTALLED_APPS={'remove': 'django.contrib.sites'})
+    @ignore_warnings(category=RemovedInDjango41Warning)
```
Comment (pope1ni) [inline tests/sitemaps_tests/test_http.py:212]:
```suggestion
    @ignore_warnings(category=RemovedInDjango50Warning)
```

File: tests/sitemaps_tests/test_generic.py [hunk 8]
```diff
@@ -57,3 +58,21 @@ def test_generic_sitemap_lastmod(self):
...
+        self.assertEqual(sitemap.get_protocol('http'), 'http')
+
+    def test_warning_raised_when_falling_back_to_default_protocol(self):
+        sitemap = GenericSitemap({"queryset": None})
+        msg = (
+            "http as the default sitemap protocol is deprecated. "
+            "https will be the new default protocol. "
+            "Set an explicit 'protocol' to silence this warning."
+        )
+        with self.assertWarnsMessage(RemovedInDjango41Warning, msg):
```
Comment (pope1ni) [inline tests/sitemaps_tests/test_generic.py:77]:
```suggestion
        with self.assertWarnsMessage(RemovedInDjango50Warning, msg):
```

File: django/contrib/sitemaps/__init__.py [hunk 9]
```diff
@@ -122,6 +124,13 @@ def location(self, item):
...
     def get_protocol(self, protocol=None):
         # Determine protocol
+        if not any((self.protocol, protocol)):
+            warnings.warn(
+                "http as the default sitemap protocol is deprecated. "
+                "https will be the new default protocol. "
+                "Set an explicit 'protocol' to silence this warning.",
+                category=RemovedInDjango41Warning
+            )
         return self.protocol or protocol or 'http'
```
Comment (pope1ni) [inline django/contrib/sitemaps/__init__.py:134]: Add a comment above this line to explain what needs to be changed after the deprecation period has expired:
```suggestion
        # RemovedInDjango50Warning: Change 'http' to 'https' after deprecation.
        return self.protocol or protocol or 'http'
```

File: django/contrib/sitemaps/__init__.py [hunk 10]
```diff
@@ -122,6 +124,13 @@ def location(self, item):
 
     def get_protocol(self, protocol=None):
         # Determine protocol
+        if not any((self.protocol, protocol)):
+            warnings.warn(
+                "http as the default sitemap protocol is deprecated. "
+                "https will be the new default protocol. "
+                "Set an explicit 'protocol' to silence this warning.",
+                category=RemovedInDjango41Warning
+            )
```
Comment (pope1ni) [inline django/contrib/sitemaps/__init__.py:133]: I think the wording could be improved a little here:

```suggestion
        if not self.protocol and not protocol:
            warnings.warn(
                "The default protocol will be changed from 'http' to 'https' "
                "in Django 5.0. Explicitly pass 'protocol' to this method or "
                "to the constructor to silence this warning.",
                category=RemovedInDjango50Warning,
            )
```

Don't forget to update the tests too.

File: tests/sitemaps_tests/test_generic.py [hunk 11]
```diff
@@ -57,3 +58,21 @@ def test_generic_sitemap_lastmod(self):
...
         self.assertXMLEqual(response.content.decode(), expected_content)
         self.assertEqual(response.headers['Last-Modified'], 'Wed, 13 Mar 2013 10:00:00 GMT')
+
+    def test_get_protocol_defined_in_constructor(self):
+        sitemap = GenericSitemap({"queryset": None}, protocol='http')
+        self.assertEqual(sitemap.get_protocol(), 'http')
+
+    def test_get_protocol_passed_as_argument(self):
+        sitemap = GenericSitemap({"queryset": None})
+        self.assertEqual(sitemap.get_protocol('http'), 'http')
```
Comment (pope1ni) [inline tests/sitemaps_tests/test_generic.py:68]: Are these tests not already covered elsewhere? If not, then they should be added first as a separate commit as they are unrelated to this deprecation.

File: django/contrib/sitemaps/__init__.py (hunk 1 above)
Comment (pope1ni) [reply django/contrib/sitemaps/__init__.py:1]: It is fine to say "Fixed" as the fellows will handle making the actual change when removing the deprecation when it matures.

**Review by pope1ni — CHANGES_REQUESTED** [review]
Hi @rohithpr 

Thanks for picking this up. Note that the deprecation should be targeted for removal in Django 5.0, not 4.1. Please can you also add a note about the deprecation to the following files:

- `docs/internals/deprecation.txt`
- `docs/releases/4.0.txt`

You can update this PR title to:

```
Fixed #32375 -- Deprecated 'http' as the default protocol for sitemaps.
```

I'll set *patch needs improvement* on ticket-32375.

**Comment by pope1ni** [comment]
We should also add a `.. deprecated:: 4.0` marker at the end of the `.. attribute:: Sitemap.protocol` section in `docs/ref/contrib/sitemaps.txt` to highlight this deprecation in the documentation:

```rst
        .. deprecated:: 4.0

           The default protocol will change from ``'http'`` to ``'https'`` in
           Django 5.0.
```

File: tests/sitemaps_tests/test_generic.py (hunk 11 above)
Comment (rohithpr) [reply tests/sitemaps_tests/test_generic.py:68]: `grep -r "get_protocol" tests/` does not return any other results. I've separated the commits now.

File: docs/internals/deprecation.txt [hunk 12]
```diff
@@ -16,6 +16,8 @@ See the :ref:`Django 4.0 release notes <deprecated-features-4.0>` for more
 details on these changes.
 
 * The ``SERIALIZE`` test setting will be removed.
+* The default sitemap 'protocol' will be changed from 'http' to 'https'
```
Comment (carltongibson) [inline docs/internals/deprecation.txt:19]: Please add a blank line between the list items.

File: docs/releases/4.0.txt [hunk 13]
```diff
@@ -441,6 +441,8 @@ Miscellaneous
 * The undocumented ``django.utils.baseconv`` module is deprecated.
 
 * The undocumented ``django.utils.datetime_safe`` module is deprecated.
+* The default sitemap 'protocol' will be changed from 'http' to 'https'
```
Comment (carltongibson) [inline docs/releases/4.0.txt:444]: Needs a blank line. But this should go in the `contrib.sitemaps` section (around line 132 at this revision.)

File: tests/sitemaps_tests/test_generic.py (hunk 11 above)
Comment (carltongibson) [inline tests/sitemaps_tests/test_generic.py:68]: For both these tests can we add a `subTest` passing both `http` and `https`, and checking the result. (As it stands there would pass if the implementation were just `return 'http'`...)

**Review by carltongibson — COMMENTED** [review]
OK, this looks good. It's essentially the pending one character change plus the warnings. 👍 

I just a few minor comments.

File: django/contrib/sitemaps/__init__.py [hunk 14]
```diff
@@ -122,6 +124,14 @@ def location(self, item):
 
     def get_protocol(self, protocol=None):
         # Determine protocol
+        if not any((self.protocol, protocol)):
```
Comment (felixxm) [inline django/contrib/sitemaps/__init__.py:127]: `any()` is nice but it's over-engineering for me, we don't need to create a tuple and use `any()` to check two values.
```suggestion
        if self.protocol is None and protocol is None:
```

File: django/contrib/sitemaps/__init__.py [hunk 15]
```diff
@@ -122,6 +124,14 @@ def location(self, item):
...
+            warnings.warn(
+                "The default sitemap protocol will be changed from 'http' to "
+                "'https' in Django 5.0. Explicitly pass 'protocol' to this "
+                "method or to the constructor to silence this warning.",
```
Comment (felixxm) [inline django/contrib/sitemaps/__init__.py:131]: I don't think that users call `get_protocol()` directly. I would rather recommend to set the `Sitemap.protocol` attribute (which is documented), e.g.
```suggestion
                "The default sitemap protocol will be changed from 'http' to "
                "'https' in Django 5.0. Set Sitemap.protocol to 'http' if you "
                "want to keep the current default behavior.",
```

File: django/contrib/sitemaps/__init__.py [hunk 16]
```diff
@@ -122,6 +124,14 @@ def location(self, item):
...
+                category=RemovedInDjango50Warning
```
Comment (felixxm) [inline django/contrib/sitemaps/__init__.py:132]: Should we increase `stacklevel`?

File: django/contrib/sitemaps/__init__.py (hunk 14 above)
Comment (pope1ni) [reply django/contrib/sitemaps/__init__.py:127]: Thanks for being more explicit about this than I was! (https://github.com/django/django/pull/14396#discussion_r632488901)

**Review by felixxm — COMMENTED** [review]
@rohithpr Thanks for updates :+1: I pushed small edits.

I think we should make it clear that the default will change only for sitemaps built outside the context of a request, so I pushed edits to docs changes.

**Comment by rohithpr** [comment]
Thanks every one for the reviews :)
